### DISCUSSION TITLE
How to create a Session "on_commit" hook?

### DISCUSSION THREAD

**Discussion by zzzeek** [sqlalchemy/sqlalchemy#6888] [Usage Questions] [answered]
How do I achieve this pattern?

```py
def verify_user(user_id: int) -> None:
    user = session.query(User).get(user_id)
    user.is_verified = True
    session.add(user)
    session.on_commit(lambda: user_verified.send(user.id))
```
Basically, when the above `Session` is committed, the given function will fire off, then it's discarded.

**Comment by zzzeek** [comment, marked as answer]
the "on_commit" recipe is implemented using the static [event hook](https://docs.sqlalchemy.org/en/14/orm/events.html?highlight=after_transaction_end#sqlalchemy.orm.SessionEvents.after_transaction_end) as a processing feature for the hooks as a stack.  The [session.info](https://docs.sqlalchemy.org/en/14/orm/session_api.html?highlight=session%20info#sqlalchemy.orm.Session.info) dictionary is used as a holding place for the stack of functions:

```py

from sqlalchemy import event

def on_commit(session, fn):
    """register a hook that will take place on next commit"""

    hooks = session.info.setdefault("on_commit_hooks", [])
    hooks.append(fn)

@event.listens_for(Session, "after_commit")
def process_hooks(session):
    """invoke hooks on commit"""

    hooks = session.info.get("on_commit_hooks", [])
    while hooks:
        fn = hooks.pop(-1)
        fn(session)

@event.listens_for(Session, "after_transaction_end")
def remove_remaining_hooks(session, transaction):
    """discard un-invoked hooks when the top level transaction ends"""

    if transaction.parent is None:
        session.info.pop("on_commit_hooks", None)


```

a full usage example is as follows:

```py
from sqlalchemy import Column
from sqlalchemy import create_engine
from sqlalchemy import event
from sqlalchemy import Integer
from sqlalchemy import String
from sqlalchemy.orm import declarative_base
from sqlalchemy.orm import Session


def on_commit(session, fn):
    """register a hook that will take place on next commit"""

    hooks = session.info.setdefault("on_commit_hooks", [])
    hooks.append(fn)

@event.listens_for(Session, "after_commit")
def process_hooks(session):
    """invoke hooks on commit"""

    hooks = session.info.get("on_commit_hooks", [])
    while hooks:
        fn = hooks.pop(-1)
        fn(session)

@event.listens_for(Session, "after_transaction_end")
def remove_remaining_hooks(session, transaction):
    """discard un-invoked hooks when the top level transaction ends"""

    if transaction.parent is None:
        session.info.pop("on_commit_hooks", None)


Base = declarative_base()
e = create_engine("sqlite://")


class A(Base):
    __tablename__ = "a"

    id = Column(Integer, primary_key=True)
    data = Column(String)


Base.metadata.create_all(e)

with Session(e) as s:
    a1 = A(data="some data")
    s.add(a1)
    on_commit(
        s,
        lambda s: print(
            f"send an email, just committed {a1.data}, id {a1.id}"
        ),
    )
    s.commit()
```

design decisions which the implementor can choose how to deal with include:

1. can one hook pre-empt the rest of the hooks?
2. what happens if a hook fails, keep running them, or stop?  propagate the exception?  etc.

Also note that many web frameworks supply existing patterns to achieve this goal, such as Pyramid's "tweens".

**Comment by reritom** [comment]
This might not be the right place to put this. But before @zzzeek posted his solution above, this had been giving me headaches for days in the context of a flask + flask_sqlalchemy project. So I will post how I adapted it for my usecase in case anyone else gets into this position (though zzzeeks solution above should be ample enough).

I adapted the solution from zzzeek for usage in flask + flask_sqlalchemy with the API  `db.session.on_commit(func)` and it seems to work as expected.

Typically in flask + flask sqlalchemy you have a module where your db object is located like this:
```py
# database.py
from flask_sqlalchemy import SQLAlchemy

db = SQLAlchemy()
```
and then in your `app.py` you import `db` and call `db.init_app(app)`.

To add the `on_commit` API, in `database.py` I did the following:
```py
# database.py
from flask_sqlalchemy import SQLAlchemy
from sqlalchemy import orm

try:
    from greenlet import getcurrent as _ident_func
except ImportError:
    from threading import get_ident as _ident_func


class EnhancedScopedSession(orm.scoped_session):
    """
    Enhance the sqlalchemy.orm.scoped_session to allow db.session.on_commit()
    https://github.com/sqlalchemy/sqlalchemy/discussions/6888
    """
    def on_commit(self, hook, *args, **kwargs):
        hooks = self.info.setdefault("on_commit_hooks", [])
        hooks.append(hook)


class EnhancedSQLAlchemy(SQLAlchemy):
    def create_scoped_session(self, options=None):
        """
        Overriding https://github.com/pallets/flask-sqlalchemy/blob/931ec00d1e27f51508e05706eef41cc4419a0b32/src/flask_sqlalchemy/__init__.py#L776
        to allow db.session.on_commit by using the EnhancedScopedSession instead of the scoped_session,
        because in practice we use the scoped_session when using db.session, and not the SignallingSession,
        so overriding "create_session" alone isn't enough.
        """  # noqa
        if options is None:
            options = {}

        scopefunc = options.pop("scopefunc", _ident_func)
        options.setdefault("query_cls", self.Query)
        return EnhancedScopedSession(self.create_session(options), scopefunc=scopefunc)

db = EnhancedSQLAlchemy()
```
and kept the two `@event.listens_for` that were defined in the zzzeeks implementation above in an `events.py` module. Hopefully there aren't any glaring problems with subclassing the scoped_session, but it seems to work. which allows me to achieve this pattern 
```py
def verify_user(user_id: int) -> None:
    user = User.query.get(user_id)  # Flask_sqlalchemy API for querying a model
    user.is_verified = True
    db.session.add(user)
    db.session.on_commit(lambda: user_verified.send(user.id))
```

**Reply by zzzeek** [reply]
you're good, glad it worked out

**Reply by reritom** [reply]
Thanks a lot though, I wasn't expecting such a comprehensive response so quickly.

**Comment by omerfarukabaci** [comment]
@zzzeek I wonder your thoughts about the following use case of this:

```py
session = ... # create a session
event.listen(session, "after_commit", process_hooks)
event.listen(session, "after_transaction_end", remove_remaining_hooks)

on_commit(session, lambda s: s.execute(text("SELECT 1;")))
session.commit()
```

This is raising:

> sqlalchemy.exc.InvalidRequestError: This session is in 'committed' state; no further SQL can be emitted within this transaction.

However, the following works fine:

```py
session = ... # create a session
event.listen(session, "after_commit", process_hooks)
event.listen(session, "after_transaction_end", remove_remaining_hooks)
session.commit()
session.execute(text("SELECT 1;"))
```

So it seems like these are not identical. I do not know the internals very well but it seems like `SessionTransactionState.COMMITTED` is not being used in anywhere except setting the state after commiting transactions, maybe further SQLs might be allowed for this state, or state can be changed to `SessionTransactionState.COMMITTED` after running `after_commit` hooks? Do you have any suggestions to be able to use the current session in on commit hooks?

**Reply by zzzeek** [reply]
your code snippet is not clear where this .execute() command is being run, but if you ran additional SQL commands in the after_commit hook, that means the session has to begin a new transaction.   this is not appropriate because we are still in the middle of the commit.    Otherwise you cause a recursive situation.   if you want to emit SQL statements in the "commit" phase of the session, use a new connection from an engine.

**Reply by omerfarukabaci** [reply]
Thank you for your quick answer @zzzeek! I see your point, but I still have some concerns:

1. I would expect the following two to be identical regardless of what `some_function` does:
    ```py
    event.listen(session, "after_commit", some_function)
    ```
    
    ```py
    session.commit()
    some_function(session)
    ```
2. If we cannot execute anything on session within the on commit hooks, why do we pass it as an argument to the on commit hook?

**Reply by zzzeek** [reply]
I'm not totally following the way you are laying out the code there.  is `some_function(session)` inside your hook, and the .commit() call is outside?   the answer to your Q in general depends on what you actually are trying to do.

The Session is passed as an argument to the event because it's the subject of the event.  there's lots you can do with the Session that does not involve starting a new transaction.

**Reply by omerfarukabaci** [reply]
I didn't know that you cannot execute queries in `after_commit` events, and this is well documented. I don't want to waste your time more, I got your point and thank you for your responses @zzzeek! 🚀 

------

To explain the situation a bit more on my side to help others:

The thing that confuses me here is that we are using `after_commit` events to implement `on_commit` hooks, but you cannot emit SQLs in `after_commit` hooks (see notes [here](https://docs.sqlalchemy.org/en/20/orm/events.html#sqlalchemy.orm.SessionEvents.after_commit))

```py
def some_function(session):
    session.execute(text("SELECT 1:"))  # `SELECT 1;` is just an arbitrary SQL, no matter what SQL this is.

session = create_session()  # creates and returns a Session object.
session.commit()
some_function(session)  # this works fine.

event.listen(session, "after_commit", some_function)
session.commit()  # this one fails
```

So AFAIU because of this you need to create a new session to be able to execute queries in an on commit hook.
